### ISSUE TITLE
Decorated functions nested inside other functions are marked as "not accessed"

### ISSUE THREAD

**Issue by evtn** [microsoft/pyright#5130]
**Describe the bug**
I'm using decorators to store some functions, and then those will be executed.
When combined with a factory pattern, it gives a warning that my function is unused (when in fact it cannot be proven)

**To Reproduce**
Minimal example would be:
```python
deco = lambda f: f

def outer():
    @deco
    def f(): # this function is "unused"
        pass
```

**Expected behavior**
This is not an error, but it takes my attention away, especially when a lot of nested functions are present. I think there should be no warning here.

**Screenshots or Code**
(example a bit simplified for clarity)
```python
store = []


def some_decorator(f):
    store.append(f)
    return f


@some_decorator
def test1():
    print("this is ok")


def outer():
    @some_decorator
    def test2():
        print("this function is flagged as not accessed")

outer()

for f in store:
    f()
```

**Comment by erictraut** [comment]
Pyright is behaving as intended here. The symbol `f` is a local symbol that is not referenced.

If you're receiving a warning, then I presume that you have enabled the `reportUnusedFunction` diagnostic check and configured it to generate a warning. By default, this is off. If you don't want to see such a warning, you can keep this diagnostic check disabled.

Pyright will generate a "hint" with a tag that tells the editor to display the text in a subtle grayed-out manner. Perhaps that's what you're seeing? That's intentional in this case.

In any event, this isn't a bug, and making the change that you're suggesting would introduce an inconsistency that is not warranted.

I'll point out that you're using a decorator in an odd manner here. If your intent is to register a function with a registration call, then I recommend that you don't use a decorator and instead simply call the registration function. This will avoid the creation of a symbol whose value is not used.

```python
def register(f):
    store.append(f)

def outer():
    def test2():
        print("this function is flagged as not accessed")

    register(test2)
```

If this suggestion doesn't appeal to you, another workaround is to make the function `_` which, by convention, means "I don't care about the name of this symbol". Pyright special cases this name as such.

**Comment by evtn** [comment]
The problem with `_` is that Pyright reports type errors on it, so I can't, for example, do this:

```python
def _():
    pass
    
value, _ = [1, 2] # Expression of type "int" cannot be assigned to declared type "() -> None"
```

Maybe this should be a separate issue?
